app/src/main/java/ch/hikemate/app/model/route/HikeRoutesRepositoryOverpass.kt

```diff
@@ -104,14 +107,102 @@ class HikeRoutesRepositoryOverpass(val client: OkHttpClient) : HikeRoutesReposit
+     * @return The list of points that represent the hike route. An empty list is returned either if
+     *   the hike is considered invalid or if there are no points in the hike.
+     */
+    private fun flattenHikeWays(hikeWays: List<HikeWay>): HikeWay {
+      val points = mutableListOf<LatLong>()
+      for (i in hikeWays.indices) {
+        if (points.isEmpty()) {
+          handleFirstWay(hikeWays, points, i)
+        } else {
+          handleSubsequentWays(hikeWays, points, i)
+        }
+      }
+      return points
+    }
+
+    /**
+     * Procedure to handles the first way of the hike during the flattening process.
+     *
+     * @param hikeWays The list of hike ways.
+     * @param points The list of points to add the way to.
+     * @param i The index of the way to handle.
+     */
+    private fun handleFirstWay(hikeWays: List<HikeWay>, points: MutableList<LatLong>, i: Int) {
+      if (i < hikeWays.size - 1 &&
+          (hikeWays[i].first() == hikeWays[i + 1].first() ||
+              hikeWays[i].first() == hikeWays[i + 1].last())) {
+        points.addAll(hikeWays[i].reversed())
+      } else {
+        points.addAll(hikeWays[i])
+      }
+    }
+
+    /**
+     * Procedure to handles the subsequent ways of the hike during the flattening process.
+     *
+     * @param hikeWays The list of hike ways.
+     * @param points The list of points to add the way to.
+     * @param i The index of the way to handle.
+     */
+    private fun handleSubsequentWays(
+        hikeWays: List<HikeWay>,
+        points: MutableList<LatLong>,
+        i: Int
+    ) {
+      when {
+        hikeWays[i].first() == points.last() ->
+            points.addAll(hikeWays[i].subList(1, hikeWays[i].size))
+        hikeWays[i].last() == points.last() ->
+            points.addAll(hikeWays[i].reversed().subList(1, hikeWays[i].size))
+        else -> handleUnconnectedWays(hikeWays, points, i)
+      }
+    }
+
+    /**
+     * Procedure to handles the unconnected ways of the hike during the flattening process.
+     *
+     * @param hikeWays The list of hike ways.
+     * @param points The list of points to add the way to.
+     * @param i The index of the way to handle.
+     */
+    private fun handleUnconnectedWays(
+        hikeWays: List<HikeWay>,
+        points: MutableList<LatLong>,
+        i: Int
+    ) {
+      val lastPoint = points.last()
+      val firstWayPoint = hikeWays[i].first()
+      val lastWayPoint = hikeWays[i].last()
+      val distanceFirst = firstWayPoint.distanceTo(lastPoint)
+      val distanceLast = lastWayPoint.distanceTo(lastPoint)
+
+      when {
+        distanceFirst < distanceLast && distanceFirst < MAX_DISTANCE_BETWEEN_OPENED_PATH ->
+            points.addAll(hikeWays[i])
+        distanceFirst >= distanceLast && distanceLast < MAX_DISTANCE_BETWEEN_OPENED_PATH ->
+            points.addAll(hikeWays[i].reversed())
+        else -> points.clear() // Discard the hikes if unconnected
```

Comment:
Warning: Does points.clear() actually discard the entire hike or just make it start at the next way in the original list of ways? I suspect the latter, since at the next iteration of the for loop in `flattenHikeWays` the method `handleFirstWay` would be called.

That would be wrong since we would potentially display hikes that start in the middle of nowhere, which might overwhelm our users, which are beginners to hiking.

Reply:
Yep nop, indeed, completely right, my refactoring broke it. I will correct that, and return a value, thus it's no longer a procedure :)